designs/DeveloperTestingPackageExperience.md

```diff
@@ -0,0 +1,125 @@
+# Developer Testing Package Experience
+
+* Status: In Review
+* Author(s): [Samuel Cragg](https://github.com/samcragg)
+* Issue: [6579](https://github.com/NuGet/Home/issues/6579) Enhance the
+  experience for the testing of packages in development
+
+## Problem Background
+
+As a NuGet package author, it would be nice to be able to test the package
+locally before publishing to a NuGet feed.
+
+As packages can include custom MsBuild targets, it's important to verify that
+the generated package works as expected when restored into a project. Being able
+to verify that locally (including on a CI server) will enable developers to spot
+changes in behaviour sooner, enhancing the inner-loop performance for .NET
+developers.
+
+## Who are the customers
+
+NuGet package authors (i.e. developers who create NuGet packages).
+
+## Goals
+
+The goal of this design is to enable a generated `nupkg` file to be used by a
+project using the existing `PackageReference` functionality. The version of the
+package may be an existing published version or that of a previously restored
+local package (i.e. there is no requirement that the version is unique, as there
+is with the current package publishing requirements), however, it is assumed
+that the package ID will be unique to avoid conflict with other dependencies.
+
+## Non-Goals
+
+Custom feeds will not be supported by the solution - the solution is only
+interested in folders containing `nupkg` files. The rationale for this is to
+keep the design simple and to emphasise that it is not a replacement for other
+package sources. Also, since the target audience is creating the packages, the
+file would have been produced in order to publish it, so this limitation is not
+perceived as an impediment.
+
+## Solution Overview
+
+In order to provide a solution, there are a few issues that need to be
+highlighted with the current machinery with regards to installing of a package
+in the context of trying to test a newly created package:
+
+* NuGet caches packages by ID and version
+* The order of multiple sources are searched through is non-deterministic
```

Comment:
https://docs.microsoft.com/en-us/nuget/concepts/package-installation-process covers this. 

File feeds are preferred over http feeds.

Reply:
True if the version is specified, however, with no version is specified then both are still checked. See #10367